numpy/lib/function_base.py

```diff
@@ -1526,19 +1530,41 @@ def unwrap(p, discont=pi, axis=-1):
     array([ 0.        ,  0.78539816,  1.57079633,  5.49778714,  6.28318531]) # may vary
     >>> np.unwrap(phase)
     array([ 0.        ,  0.78539816,  1.57079633, -0.78539816,  0.        ]) # may vary
-
+    >>> unwrap([0, 1, 2, -1, 0], period=4)
+    array([0., 1., 2., 3., 4.])
+    >>> unwrap([ 1, 2, 3, 4, 5, 6, 1, 2, 3], period=6)
+    array([1., 2., 3., 4., 5., 6., 7., 8., 9.])
+    >>> unwrap([2, 3, 4, 5, 2, 3, 4, 5], period=4)
+    array([2., 3., 4., 5., 6., 7., 8., 9.])
+    >>> phase_deg = np.mod(np.linspace(0,720,19), 360) - 180
+    >>> unwrap(phase_deg, period=360)
+    array([-180., -140., -100.,  -60.,  -20.,   20.,   60.,  100.,  140.,
+        180.,  220.,  260.,  300.,  340.,  380.,  420.,  460.,  500.,
+        540.])
     """
     p = asarray(p)
     nd = p.ndim
     dd = diff(p, axis=axis)
+    if discont is None:
+        discont = period/2
     slice1 = [slice(None, None)]*nd     # full slices
     slice1[axis] = slice(1, None)
     slice1 = tuple(slice1)
-    ddmod = mod(dd + pi, 2*pi) - pi
-    _nx.copyto(ddmod, pi, where=(ddmod == -pi) & (dd > 0))
+    dtype = np.result_type(dd, period)
+    if _nx.issubdtype(dtype, _nx.integer):
+        interval_low = -(period // 2)
+        interval_high = -interval_low
+    else:
+        interval_low = -period / 2
+        interval_high = -interval_low
+    ddmod = mod(dd - interval_low, period) + interval_low
+    if period % 2 == 0:
+        # for `mask = (abs(dd) == period/2)`, the above line made `ddmod[mask] == -period/2`.
+        # correct these such that `ddmod[mask] == sign(dd[mask])*period/2`.
+        _nx.copyto(ddmod, interval_high, where=(ddmod == interval_low) & (dd > 0))
```

Comment:
I've just realized that if we had #9963 then all of this (including the integer special case) would go away

Reply:
```python
In [199]: [math.remainder(i, 4) for i in [-4, -3, -2, -1, 0, 1, 2, 3, 4]]
Out[199]: [-0.0, 1.0, -2.0, -1.0, 0.0, 1.0, 2.0, -1.0, 0.0]
```